Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,931 @@
+    using TauTuple = tau_tuple::TauTuple;
+
+    DataLoader() :
+        n_tau(setup::n_tau), current_entry(setup::start_dataset), start_dataset(setup::start_dataset),
+        n_inner_cells(setup::n_inner_cells), inner_cell_size(setup::inner_cell_size), n_outer_cells(setup::n_outer_cells),
+        outer_cell_size(setup::outer_cell_size), n_threads(setup::n_threads),
+        n_fe_tau(setup::n_fe_tau), n_pf_el(setup::n_pf_el) ,n_pf_mu(setup::n_pf_mu),
+        n_pf_chHad(setup::n_pf_chHad), n_pf_nHad(setup::n_pf_nHad) ,n_pf_gamma(setup::n_pf_gamma),
+        n_electrons(setup::n_ele), n_muons(setup::n_muon), n_labels(setup::tau_types), labels_names(setup::tau_types_names),
+        n_eta_bins(setup::n_eta_bins), eta_min(setup::eta_min), eta_max(setup::eta_max),
+        n_pt_bins(setup::n_pt_bins), pt_min(setup::pt_min), pt_max(setup::pt_max),
+        innerCellGridRef(n_inner_cells, n_inner_cells, inner_cell_size, inner_cell_size),
+        outerCellGridRef(n_outer_cells, n_outer_cells, outer_cell_size, outer_cell_size),
+        input_files(FindInputFiles(setup::input_dirs,setup::file_name_pattern,
+                                   setup::exclude_list, setup::exclude_dir_list))
+    {
+      if(n_threads > 1) ROOT::EnableImplicitMT(n_threads);
+
+      // file = OpenRootFile(file_name);
+      // tauTuple = std::make_shared<tau_tuple::TauTuple>(file.get(), true);
+
+      std::cout << "Number of files to process: " << input_files.size() << std::endl;
+      tauTuple = std::make_shared<TauTuple>("taus", input_files, true);
+      end_dataset = std::min(setup::end_dataset, tauTuple->GetEntries());
+
+      // histogram to calculate weights
+      // TH1::AddDirectory(kFALSE); // sets a global switch disabling the referencing
+      TFile* file_input = new TFile(setup::input_spectrum.c_str());
+      TFile* file_target = new TFile(setup::target_spectrum.c_str());
+      TH2D* target_hist = (TH2D*)file_target->Get("eta_pt_hist_tau");
+
+      for(int type = 0; type < 4; ++type) {
+        hist_weights.push_back(shared_ptr<TH2D>(new TH2D(("w_1_"+labels_names[type]).c_str(),
+                                                         ("w_1_"+labels_names[type]).c_str(),
+                                                         n_eta_bins, eta_min, eta_max,
+                                                         n_pt_bins, pt_min, pt_max)));
+        hist_weights[type]->SetDirectory(0);
+
+        TH2D* after_rebin_input_hist = new TH2D("input_hist", "input_hist",
+                                                n_eta_bins, eta_min, eta_max,
+                                                n_pt_bins, pt_min, pt_max);
+        TH2D* input_hist = (TH2D*)file_input->Get(("eta_pt_hist_"+labels_names[type]).c_str());
+
+        RebinAndFill(*hist_weights[type], *target_hist);
+        RebinAndFill(*after_rebin_input_hist, *input_hist);
+        hist_weights[type]->Divide(after_rebin_input_hist);
+        delete after_rebin_input_hist;
+      }
+
```

Comment:
Implement a control that `max_weight / min_weight < setup::weight_thr`. If not satisfied -> throw an exception.